File: packages/scanner/src/mixed-content.ts
import type { Finding } from "./types.js";

export interface MixedContentAsset {
  tag: string;
  attr: string;
  url: string;
}

export function detectMixedContent(htmlBody: string, isHttps: boolean): Finding[] {
  const findings: Finding[] = [];

  if (!isHttps || !htmlBody) {
    findings.push({
      id: "mixed-content-skipped",
      category: "mixed_content",
      categoryTitle: "Contenu mixte",
      title: "Contrôle de contenu mixte non applicable",
      severity: "info",
      status: "pass",
      description: "Le site n'est pas desservi en HTTPS ou le contenu HTML est indisponible.",
      importance: "Le contenu mixte ne concerne que les pages sécurisées chargeant des sous-ressources non chiffrées.",
      recommendation: "Activez HTTPS avant d'évaluer le contenu mixte.",
    });
    return findings;
  }

  const insecureAssets: MixedContentAsset[] = [];

  // Regex patterns for common tags loading subresources with http://
  const tagPatterns: Array<{ tag: string; attr: string; regex: RegExp }> = [
    { tag: "script", attr: "src", regex: /<script[^>]+src=["'](http:\/\/[^"']+)["']/gi },
    { tag: "link", attr: "href", regex: /<link[^>]+href=["'](http:\/\/[^"']+)["']/gi },
    { tag: "img", attr: "src", regex: /<img[^>]+src=["'](http:\/\/[^"']+)["']/gi },
    { tag: "iframe", attr: "src", regex: /<iframe[^>]+src=["'](http:\/\/[^"']+)["']/gi },
    { tag: "audio", attr: "src", regex: /<audio[^>]+src=["'](http:\/\/[^"']+)["']/gi },
    { tag: "video", attr: "src", regex: /<video[^>]+src=["'](http:\/\/[^"']+)["']/gi },
  ];

  for (const { tag, attr, regex } of tagPatterns) {
    let match: RegExpExecArray | null;
    while ((match = regex.exec(htmlBody)) !== null) {
      if (match[1]) {
        insecureAssets.push({ tag, attr, url: match[1] });
      }
    }
  }

  if (insecureAssets.length > 0) {
    const sampleUrls = insecureAssets.slice(0, 3).map((a) => `<${a.tag} ${a.attr}="${a.url}">`).join("\n");
    const hasActiveMixed = insecureAssets.some((a) => a.tag === "script" || a.tag === "iframe");

    findings.push({
      id: "mixed-content-detected",
      category: "mixed_content",
      categoryTitle: "Contenu mixte",
      title: `${insecureAssets.length} ressource(s) HTTP non chiffrée(s) détectée(s) sur page HTTPS`,
      severity: hasActiveMixed ? "high" : "medium",
      status: "fail",
      description: `La page HTTPS charge ${insecureAssets.length} ressource(s) non sécurisée(s) via le protocole http://.`,
      importance: "Les navigateurs bloquent souvent le contenu mixte actif (scripts, iframes) ou dégradent l'indicateur de cadenas pour le contenu passif (images), avertissant les internautes d'une insécurité.",
      recommendation: "Remplacez toutes les URL 'http://' par des URL relatives ou 'https://', ou configurez la directive CSP 'upgrade-insecure-requests'.",
      remediationSnippet: "Content-Security-Policy: upgrade-insecure-requests;",
      detectedValue: sampleUrls,
      cwe: "CWE-311",
    });
  } else {
    findings.push({
      id: "mixed-content-pass",
      category: "mixed_content",
      categoryTitle: "Contenu mixte",
      title: "Aucun contenu mixte détecté",
      severity: "low",
      status: "pass",
      description: "Toutes les ressources observables (scripts, feuilles de styles, images, iframes) utilisent des liaisons chiffrées ou relatives.",
      importance: "Préserve l'intégrité de l'expérience HTTPS sans avertissement de cadenas cassé.",
      recommendation: "Continuez d'imposer des liens HTTPS pour toutes les futures ressources tierces.",
      detectedValue: "0 ressource non chiffrée",
    });
  }

  return findings;
}

File: packages/python/securio/checks/mixed_content.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import List
from ..models import Finding

CATEGORY_KEY = "mixed_content"
CATEGORY_TITLE = "Contenu mixte"


@dataclass
class MixedContentAsset:
    tag: str
    attr: str
    url: str


TAG_PATTERNS = [
    ("script", "src", re.compile(r"""<script[^>]+src=["'](http://[^"']+)["']""", re.IGNORECASE)),
    ("link", "href", re.compile(r"""<link[^>]+href=["'](http://[^"']+)["']""", re.IGNORECASE)),
    ("img", "src", re.compile(r"""<img[^>]+src=["'](http://[^"']+)["']""", re.IGNORECASE)),
    ("iframe", "src", re.compile(r"""<iframe[^>]+src=["'](http://[^"']+)["']""", re.IGNORECASE)),
    ("audio", "src", re.compile(r"""<audio[^>]+src=["'](http://[^"']+)["']""", re.IGNORECASE)),
    ("video", "src", re.compile(r"""<video[^>]+src=["'](http://[^"']+)["']""", re.IGNORECASE)),
]


def detect_mixed_content(html_body: str, is_https: bool) -> List[Finding]:
    findings: List[Finding] = []

    if not is_https or not html_body:
        findings.append(
            Finding(
                id="mixed-content-skipped",
                category=CATEGORY_KEY,
                category_title=CATEGORY_TITLE,
                title="Contrôle de contenu mixte non applicable",
                severity="info",
                status="pass",
                description="Le site n'est pas desservi en HTTPS ou le contenu HTML est indisponible.",
                importance="Le contenu mixte ne concerne que les pages sécurisées chargeant des sous-ressources non chiffrées.",
                recommendation="Activez HTTPS avant d'évaluer le contenu mixte.",
            )
        )
        return findings

    insecure_assets: List[MixedContentAsset] = []
    for tag, attr, pattern in TAG_PATTERNS:
        for match in pattern.finditer(html_body):
            url = match.group(1)
            insecure_assets.append(MixedContentAsset(tag=tag, attr=attr, url=url))

    if insecure_assets:
        sample_urls = "\n".join(
            f"<{a.tag} {a.attr}=\"{a.url}\">" for a in insecure_assets[:3]
        )
        has_active_mixed = any(a.tag in ("script", "iframe") for a in insecure_assets)

        findings.append(
            Finding(
                id="mixed-content-detected",
                category=CATEGORY_KEY,
                category_title=CATEGORY_TITLE,
                title=f"{len(insecure_assets)} ressource(s) HTTP non chiffrée(s) détectée(s) sur page HTTPS",
                severity="high" if has_active_mixed else "medium",
                status="fail",
                description=f"La page HTTPS charge {len(insecure_assets)} ressource(s) non sécurisée(s) via le protocole http://.",
                importance=(
                    "Les navigateurs bloquent souvent le contenu mixte actif (scripts, iframes) "
                    "ou dégradent l'indicateur de cadenas pour le contenu passif (images), avertissant les internautes d'une insécurité."
                ),
                recommendation=(
                    "Remplacez toutes les URL 'http://' par des URL relatives ou 'https://', "
                    "ou configurez la directive CSP 'upgrade-insecure-requests'."
                ),
                remediation_snippet="Content-Security-Policy: upgrade-insecure-requests;",
                detected_value=sample_urls,
                cwe="CWE-311",
            )
        )
    else:
        findings.append(
            Finding(
                id="mixed-content-pass",
                category=CATEGORY_KEY,
                category_title=CATEGORY_TITLE,
                title="Aucun contenu mixte détecté",
                severity="low",
                status="pass",
                description="Toutes les ressources observables (scripts, feuilles de styles, images, iframes) utilisent des liaisons chiffrées ou relatives.",
                importance="Préserve l'intégrité de l'expérience HTTPS sans avertissement de cadenas cassé.",
                recommendation="Continuez d'imposer des liens HTTPS pour toutes les futures ressources tierces.",
                detected_value="0 ressource non chiffrée",
            )
        )

    return findings
